File: leetcode/python/000417.pacific-atlantic-water-flow.py
from collections import deque

class Solution:
    # Time complexity: O(m * n)
    # Space complexity: O(m * n)
    def pacificAtlantic(self, heights: list[list[int]]) -> list[list[int]]:
        m, n = len(heights), len(heights[0])
        p_queue, a_queue = deque(), deque()
        p_seen, a_seen = set(), set()

        for i in range(n):
            p_queue.append((0, i))
            p_seen.add((0, i))

        for i in range(1, m):
            p_queue.append((i, 0))
            p_seen.add((i, 0))

        for i in range(m):
            a_queue.append((i, n - 1))
            a_seen.add((i, n - 1))

        for i in range(n - 1):
            a_queue.append((m - 1, i))
            a_seen.add((m - 1, i))

        def bfs(queue: deque[tuple[int, int]], seen: set[tuple[int, int]]):
            while queue:
                i, j = queue.popleft()
                for i_off, j_off in [(0, 1), (0, -1), (-1, 0), (1, 0)]:
                    r, c = i + i_off, j + j_off
                    if 0 <= r < m and 0 <= c < n and (r, c) not in seen and heights[r][c] >= heights[i][j]:
                        seen.add((r, c))
                        queue.append((r, c))

        bfs(p_queue, p_seen)
        bfs(a_queue, a_seen)
        return list(p_seen.intersection(a_seen))

File: leetcode/javascript/000417.pacific-atlantic-water-flow.js
/**
 * @param {number[][]} heights
 * @return {number[][]}
 * Time complexity: O(m * n)
 * Space complexity: O(m * n)
 */
var pacificAtlantic = function (heights) {
  const rows = heights.length;
  const cols = heights[0].length;
  const pQueue = new Queue();
  const aQueue = new Queue();
  const pSeen = new Set();
  const aSeen = new Set();

  function toSetKey(r, c) {
    return `${r},${c}`;
  }

  function fromSetKey(key) {
    return key.split(",").map(Number);
  }

  for (let c = 0; c < cols; c++) {
    pQueue.enqueue([0, c]);
    pSeen.add(toSetKey(0, c));
  }

  for (let r = 1; r < rows; r++) {
    pQueue.enqueue([r, 0]);
    pSeen.add(toSetKey(r, 0));
  }

  for (let c = 0; c < cols; c++) {
    aQueue.enqueue([rows - 1, c]);
    aSeen.add(toSetKey(rows - 1, c));
  }

  for (let r = 0; r < rows - 1; r++) {
    aQueue.enqueue([r, cols - 1]);
    aSeen.add(toSetKey(r, cols - 1));
  }

  function bfs(queue, seen) {
    while (!queue.isEmpty()) {
      const [r, c] = queue.dequeue();

      for (const [r_off, c_off] of [[-1, 0], [1, 0], [0, 1], [0, -1]]) {
        const newR = r + r_off;
        const newC = c + c_off;

        if (
          newR >= 0 && newC >= 0 && newR < rows && newC < cols &&
          !seen.has(toSetKey(newR, newC)) &&
          heights[newR][newC] >= heights[r][c]
        ) {
          queue.enqueue([newR, newC]);
          seen.add(toSetKey(newR, newC));
        }
      }
    }
  }

  bfs(pQueue, pSeen);
  bfs(aQueue, aSeen);

  return Array.from(pSeen.intersection(aSeen)).map((setKey) =>
    fromSetKey(setKey)
  );
};

// Cleaned up
var pacificAtlantic2 = function (heights) {
  const rows = heights.length;
  const cols = heights[0].length;
  const pSeen = new Set();
  const aSeen = new Set();
  const pQueue = new Queue();
  const aQueue = new Queue();

  function toSetKey(r, c) {
    return `${r},${c}`;
  }

  function fromSetKey(key) {
    return key.split(",").map(Number);
  }

  for (let c = 0; c < cols; c++) {
    pQueue.enqueue([0, c]);
    pSeen.add(toSetKey(0, c));
    aQueue.enqueue([rows - 1, c]);
    aSeen.add(toSetKey(rows - 1, c));
  }

  for (let r = 0; r < rows; r++) {
    pQueue.enqueue([r, 0]);
    pSeen.add(toSetKey(r, 0));
    aQueue.enqueue([r, cols - 1]);
    aSeen.add(toSetKey(r, cols - 1));
  }

  function bfs(queue, seen) {
    while (!queue.isEmpty()) {
      const [r, c] = queue.dequeue();

      for (const [offR, offC] of [[-1, 0], [1, 0], [0, 1], [0, -1]]) {
        const newR = r + offR;
        const newC = c + offC;

        if (
          newR >= 0 && newC >= 0 && newR < rows && newC < cols &&
          !seen.has(toSetKey(newR, newC)) &&
          heights[newR][newC] >= heights[r][c]
        ) {
          queue.enqueue([newR, newC]);
          seen.add(toSetKey(newR, newC));
        }
      }
    }
  }

  bfs(pQueue, pSeen);
  bfs(aQueue, aSeen);

  return Array.from(pSeen.intersection(aSeen)).map(fromSetKey);
};
